version: add major, minor, patch methods

SeekingMeaning opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This adds five new methods to the `Version` class: `major`, `minor`, `patch`, `major_minor`, `major_minor_patch`. In particular, `Version#major` will be useful for determining the version suffix of `gcc`: Homebrew/homebrew-core#59485

(Based on [`cask/dsl/version`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cask/dsl/version.rb))

### Example usage

```rb
version_suffix = Formula["gcc"].linked_version.version.major
```

Examples in `brew irb`:

```
irb(main):001:0> Version.new("10.2")
=> #<Version:0x00007fa606210e90 @version="10.2">
```

```
irb(main):002:0> Version.new("10.2").major
=> #<Version::NumericToken 10>
```

```
irb(main):003:0> Version.new("10.2").minor
=> #<Version::NumericToken 2>
```

```
irb(main):004:0> Version.new("10.2").patch
=> nil
```

```
irb(main):005:0> Version.new("10.2").major_minor
=> #<Version:0x00007fb489879e30 @version="10.2">
```

```
irb(main):006:0> Version.new("10.2").major_minor_patch
=> #<Version:0x00007fb489885c08 @version="10.2">
```

SeekingMeaning commented:
One thing I noticed is that `Version::NumericToken`s can't be compared with strings or integers:

```
irb(main):001:0> Version::NumericToken.new("1") == "1"
=> false
irb(main):002:0> Version::NumericToken.new("1") == 1
=> false
```

Whereas this works fine with `Version`:

```
irb(main):003:0> Version.new("1") == 1
=> true
irb(main):004:0> Version.new("1") == "1"
=> true
```

I think I'll open another PR to address this

reitermarkus commented:
Might be good to make this a module so it can be shared between both `Cask::Version` and `Version`.

SeekingMeaning commented:
That would be a bit complex since `Cask::DSL::Version` is subclass of `String`:

```rb
module Cask
  class DSL
    class Version < ::String
```

Would it perhaps make sense to make `Version` also a subclass of `String`?

MikeMcQuaid commented:
> In particular, `Version#major` will be useful for determining the version suffix of `gcc`, such as in [Homebrew/homebrew-core#58973](https://github.com/Homebrew/homebrew-core/pull/58973)

Can you provide an example of how it should be used?



> Might be good to make this a module so it can be shared between both `Cask::Version` and `Version`.

Agreed with this, I'd really love to see this deduplicate some Cask code.

dtrodrigues commented:
Can a `formula_suffix` method be added to be used in `audit.rb`? The existing changes seem like they would allow the minor version parsing to be removed from `audit.rb`:

https://github.com/Homebrew/brew/blob/c49ef80dd37bcdbbb2d33b321e58a46fdd102267/Library/Homebrew/dev-cmd/audit.rb#L754-L760

gromgit commented:
> Can you provide an example of how it should be used?

It would turn the following incantation in the `augustus` formula:
```
with_env("HOMEBREW_CC" => Formula["gcc"].opt_bin/"gcc-#{Formula["gcc"].linked_version.to_s.slice(/\d+/)}") do
```
into:
```
with_env("HOMEBREW_CC" => Formula["gcc"].opt_bin/"gcc-#{Formula["gcc"].linked_version.version.major}") do
```
Not that much shorter, but somewhat more comprehensible to the Ruby-challenged like me.

There are at least 5 other formulae I'm tracking that could use the same improvement.

MikeMcQuaid commented:
> Not that much shorter, but somewhat more comprehensible to the Ruby-challenged like me.

Yup, makes sense. Can this be stripped down to the minimum number of methods we definitely have a use for in formulae and/or Homebrew/brew (including cask)?

SeekingMeaning commented:
> Can a `formula_suffix` method be added to be used in `audit.rb`? The existing changes seem like they would allow the minor version parsing to be removed from `audit.rb`:
> 
> https://github.com/Homebrew/brew/blob/c49ef80dd37bcdbbb2d33b321e58a46fdd102267/Library/Homebrew/dev-cmd/audit.rb#L754-L760

Assuming `stable_version_string.split(".").last.to_i` always refers to the `patch` version (i.e. `major.minor.patch`), we could do something like this:

```rb
      stable_version_string = stable.version.to_s
      stable_url_version = Version.parse(stable.url)
      stable_url_minor_version = stable_url_version.minor.to_i

      formula_suffix = stable.version.patch.to_i
```

Also a few lines down we could replace `stable_version_string.split(".")[0..1].join(".")` with `stable.version.major_minor.to_s`

SeekingMeaning commented:
> Can this be stripped down to the minimum number of methods we definitely have a use for in formulae and/or Homebrew/brew (including cask)?

I think all of these methods except maybe `minor_patch` will have a use <!-- (but I'm still leaning towards keeping `minor_patch`) -->

- `major`, `minor`, `patch` - used by the other methods
- `major` specifically - can be useful for determining GCC prefix as Adrian mentioned above
- `major_minor` - could be useful for includes such as `-I#{include}/gupnp-1.2` where `1.2` can be replaced with `Formula["gupnp"].version.major_minor`
<!-- - `major_minor_patch` - might be useful the same way as `major_minor` but for versions in the style of `major.minor.patch.build` -->

> > Might be good to make this a module so it can be shared between both `Cask::Version` and `Version`.
> 
> Agreed with this, I'd really love to see this deduplicate some Cask code.

I tried doing this but I don't think there's really any shared logic in how `cask/dsl/version` and `version` handle version tokens other than having the same method names

MikeMcQuaid commented:
> I think all of these methods except maybe `minor_patch` will have a use

Let's skip this for now and otherwise can include them.

> While we might be able to merge the two, we would have to do a lot of wrangling and be very careful to not introduce any regressions or unexpected behavior

This would be a great project to attempt! A general good start should be seeing if you can use our version logic without changing the tests.

dtrodrigues commented:
>Assuming `stable_version_string.split(".").last.to_i` always refers to the `patch` version (i.e. `major.minor.patch`), we could do something like this:

For some reason, I thought that it was almost always the `patch` version but not always. Looking through the existing throttled formulae though it does seem like it's always the `patch`, so just using `patch` would work and there isn't a need for an additional `suffix` method.

SeekingMeaning commented:
Made a number of quality-of-life additions/changes, hopefully nothing will break

MikeMcQuaid commented on Library/Homebrew/version.rb:
```diff
@@ -469,6 +486,26 @@ def <=>(other)
   alias eql? ==
 
+  def major
+    tokens[0]
```
```suggestion
    tokens.first
```

MikeMcQuaid commented on Library/Homebrew/version.rb:
```diff
@@ -469,6 +486,26 @@ def <=>(other)
+  end
+
+  def minor
+    tokens[1]
```
```suggestion
    tokens.second
```

MikeMcQuaid commented on Library/Homebrew/version.rb:
```diff
@@ -469,6 +486,26 @@ def <=>(other)
+  end
+
+  def patch
+    tokens[2]
```
```suggestion
    tokens.third
```

MikeMcQuaid commented on Library/Homebrew/version.rb:
```diff
@@ -469,6 +486,26 @@ def <=>(other)
+  end
+
+  def major_minor
+    Version.new([major, minor].reject(&:nil?).join("."))
```
```suggestion
    Version.new([major, minor].compact.join("."))
```

MikeMcQuaid commented on Library/Homebrew/version.rb:
```diff
@@ -469,6 +486,26 @@ def <=>(other)
+  end
+
+  def major_minor_patch
+    Version.new([major, minor, patch].reject(&:nil?).join("."))
```
```suggestion
    Version.new([major, minor, patch].compact.join("."))
```
